### ISSUE TITLE
IXmlSerializable makes XmlSerializer Deserialize creates instance of base class instead of derived class

### ISSUE THREAD

**Issue by MMariusch** [dotnet/runtime#101654]
### Description

When a base class implements IXmlSerializable then XmlSerializer Deserializing tries to create an instance of base class and calls it's virtual ReadXml instead of the overriden  ReadXml in the derived class. 

### Reproduction Steps

In the following example the deserializer is trying to make an instance of the base class instead of the derived one.

public class TestClass
{
	[XmlElement("AClass", typeof(AClass))]
	List<BaseClass> objects = new List<BaseClass>();
	
	public TestClass(){}
	
	private void SaveToFile()
	{
		objects.Add(new AClass("someString"));
		objects.Add(new AClass("anotherString"));
		var xmlSerializer = new XmlSerializer(typeof(List<BaseClass>));
		using (var writer = new StreamWriter("SaveList.xml"))
		{
			xmlSerializer.Serialize(writer, objects);
		}
	}

	private void LoadFromFile()
	{
		if (File.Exists("SaveList.xml"))
		{
			using (var reader = new StreamReader("SaveList.xml"))
			{
				var deserializedList = new XmlSerializer(typeof(List<BaseClass>)).Deserialize(reader) as List<BaseClass>;
				if (deserializedList != null && deserializedList.Count > 0)
				{
					objects = deserializedList;
				}
			}
		}
	}
}

[XmlInclude(typeof(AClass))]
public class BaseClass : IXmlSerializable
{
	public BaseClass() { }
	public XmlSchema GetSchema() { return null; }
	public virtual void ReadXml(XmlReader reader) 
	{
		Console.Write("It shouldn't be triggered.");
	}
	public virtual void WriteXml(XmlWriter writer) { }
}

public class AClass : BaseClass
{
	private string _stringVar;
	public string StringVar { get => _stringVar; private set => _stringVar = value; }
	public AClass() { }
	public AClass(string stringVar)
	{
		_stringVar = stringVar;
	}

	public override void ReadXml(XmlReader reader)
	{
		reader.MoveToContent();
		var anyElements = !reader.IsEmptyElement;
		reader.ReadStartElement();
		if (anyElements)
		{
			_stringVar = reader.ReadElementContentAsString("StringVar", "");
			reader.ReadEndElement();
		}
	}

	public override void WriteXml(XmlWriter writer)
	{
		writer.WriteAttributeString("xsi", "type", null, "AClass");
		writer.WriteElementString("StringVar", _stringVar);
	}
}

### Expected behavior

Should create an instance of the derived class.

### Actual behavior

Creates an instance of the base class.

**Comment by MMariusch** [comment]
Thank you for the answer. If I understand correctly, if I want to add another derived class then I need to modify the scheme. 
I made something like this:
`internal static XmlSchema Schema { get; } = new XmlSchema()
		{
			Items = { new XmlSchemaComplexType() { Name = "BaseClass" },
					  new XmlSchemaComplexType() { Name = "AClass", ContentModel = new XmlSchemaComplexContent() { Content = new XmlSchemaComplexContentExtension() {
						  BaseTypeName = new XmlQualifiedName("BaseClass"), Attributes = { new XmlSchemaAttribute() { Name = "AStringVar", SchemaTypeName = new XmlQualifiedName("string", ns: XmlSchema.Namespace), }, },},},},
					  new XmlSchemaComplexType() { Name = "BClass", ContentModel = new XmlSchemaComplexContent() { Content = new XmlSchemaComplexContentExtension() {
						  BaseTypeName = new XmlQualifiedName("BaseClass"), Attributes = { new XmlSchemaAttribute() { Name = "BStringVar", SchemaTypeName = new XmlQualifiedName("string", ns: XmlSchema.Namespace), }, },},},},
			}
		};`

Unfortunately this doesn't work and the deserializer skips BClass.

**Comment by MMariusch** [comment]
Thank you for the answer. If I understand correctly, if I want to add another derived class then I need to modify the scheme. 
I made something like this:

```cs
internal static XmlSchema Schema { get; } = new XmlSchema()
{
  Items = 
  { 
    new XmlSchemaComplexType() { Name = "BaseClass" },
    new XmlSchemaComplexType() 
    {
      Name = "AClass", ContentModel = new XmlSchemaComplexContent()
      {
        Content = new XmlSchemaComplexContentExtension()
        {
          BaseTypeName = new XmlQualifiedName("BaseClass"), Attributes =
          {
            new XmlSchemaAttribute() 
            { Name = "AStringVar", SchemaTypeName = new XmlQualifiedName("string", ns: XmlSchema.Namespace), },
          },
        },
      },
    },
    new XmlSchemaComplexType() 
    { 
      Name = "BClass", ContentModel = new XmlSchemaComplexContent()
      { 
        Content = new XmlSchemaComplexContentExtension() 
        {
          BaseTypeName = new XmlQualifiedName("BaseClass"), Attributes = 
          { 
            new XmlSchemaAttribute()
            { Name = "BStringVar", SchemaTypeName = new XmlQualifiedName("string", ns: XmlSchema.Namespace), }, 
          },
        },
      },
    },
  }
};
```

Unfortunately this doesn't work. The deserializer skips BClass.

**Comment by MMariusch** [comment]
Oops. I forgot to add the XmlInclude attribute. Everything works fine. Thank you. 

It's quite a lot of work to make IXmlSerializable work. So I wondering if this isn't a bug or something.

**Comment by MMariusch** [comment]
You are right. An improved documentation would be also really helpful. But personally I would prefer to keep modifications of the base class to a minimum.

Currently I've encountered another problem. I changed the BaseClass into an abstract class and updated the XmlSchemaComplexType of "BaseClass" by setting IsAbstract flag to true. Unfortunately I get an error because deserializer tries to create an instance of BaseClass instead of the derived class.

I updated my project on repo: https://github.com/MMariusch/Example

**Comment by Jrpalazzo** [comment]
@MMariusch Just to contribute for future readers. If you remove the constructor from the abstract BaseClass class it deserializes correctly.

**Comment by MMariusch** [comment]
Thank you for your suggestion. It does work, but I think a better solution would be to make the constructor protected rather than removing it.

Also, this is still a workaround because adding "IsAbstract = true" to "XmlSchemaComplexType" should be enough (if I understand correctly).
